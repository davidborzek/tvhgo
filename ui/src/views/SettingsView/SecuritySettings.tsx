import { useFormik } from 'formik';
import { useRef } from 'react';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';
import useFormikErrorFocus from '../../hooks/formik';
import { useUpdateUserPassword } from '../../hooks/user';
import Input from '../../components/Input/Input';
import Button from '../../components/Button/Button';

import styles from './SettingsView.module.scss';
import Form from '../../components/Form/Form';
import { useManageSessions } from '../../hooks/session';
import SessionList from '../../components/SessionList/SessionList';

const SecuritySettings = () => {
  const { t } = useTranslation();
  const { sessions, error, revokeSession } = useManageSessions();
  const { updatePassword } = useUpdateUserPassword();

  const currentPasswordRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const passwordRepeatRef = useRef<HTMLInputElement>(null);

  const passwordChangeValidationSchema = Yup.object().shape({
    currentPassword: Yup.string().required(t('input_required') || ''),
    password: Yup.string()
      .required(t('input_required') || '')
      .min(8, t('password_min_chars_error') || ''),
    passwordRepeat: Yup.string()
      .required(t('passwords_do_not_match') || '')
      .oneOf([Yup.ref('password')], t('passwords_do_not_match') || ''),
  });

  const passwordChangeFormik = useFormik({
    initialValues: {
      currentPassword: '',
      password: '',
      passwordRepeat: '',
    },
    validationSchema: passwordChangeValidationSchema,
    onSubmit: async ({ password, currentPassword }) => {
      updatePassword({ password, currentPassword }).then((success) => {
        // TODO: can this be done better
        currentPasswordRef.current?.blur();
        passwordRef.current?.blur();
        passwordRepeatRef.current?.blur();
        success && passwordChangeFormik.resetForm();
      });
    },
  });

  useFormikErrorFocus(
    passwordChangeFormik,
    currentPasswordRef,
    passwordRef,
    passwordRepeatRef
  );

  return (
    <>
      <div className={styles.row}>
        <Form
          onSubmit={passwordChangeFormik.handleSubmit}
          className={styles.section}
        >
          <Input
            placeholder={t('current_password')}
            label={t('current_password')}
            value={passwordChangeFormik.values.currentPassword}
            name="currentPassword"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.currentPassword
                ? passwordChangeFormik.errors.currentPassword
                : undefined
            }
            ref={currentPasswordRef}
            fullWidth
          />
          <Input
            placeholder={t('password')}
            label={t('password')}
            value={passwordChangeFormik.values.password}
            name="password"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.password
                ? passwordChangeFormik.errors.password
                : undefined
            }
            ref={passwordRef}
            fullWidth
          />
          <Input
            label={t('password_repeat')}
            placeholder={t('password_repeat')}
            value={passwordChangeFormik.values.passwordRepeat}
            name="passwordRepeat"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.passwordRepeat
                ? passwordChangeFormik.errors.passwordRepeat
                : undefined
            }
            ref={passwordRepeatRef}
            fullWidth
          />
          <div>
            <Button type="submit" label={t('save')} />
          </div>
        </Form>
      </div>
      <div className={styles.row}>
        <SessionList sessions={sessions} onRevoke={revokeSession} />
      </div>
    </>
  );
};

export default SecuritySettings;
