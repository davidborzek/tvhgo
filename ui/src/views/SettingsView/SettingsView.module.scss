.Settings {
  padding: 2rem;
  overflow: auto;
  height: 100vh;
  display: flex;
  flex-direction: column;
  gap: 2rem;
}
