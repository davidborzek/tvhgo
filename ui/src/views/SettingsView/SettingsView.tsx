import { useFormik } from 'formik';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';
import Button from '../../components/Button/Button';
import Dropdown, { Option } from '../../components/Dropdown/Dropdown';
import Form from '../../components/Form/Form';
import Input from '../../components/Input/Input';
import { Theme, useTheme } from '../../contexts/ThemeContext';
import { useUpdateUser } from '../../hooks/user';
import i18n from '../../i18n/i18n';
import styles from './SettingsView.module.scss';
import * as Yup from 'yup';
import { useEffect, useRef } from 'react';
import useFormikErrorFocus from '../../hooks/formik';
import { useAuth } from '../../contexts/AuthContext';

function SettingsView() {
  const { user } = useAuth();

  const navigate = useNavigate();
  const { t } = useTranslation();

  const { setTheme, theme } = useTheme();

  const { update } = useUpdateUser();

  const themeOptions: Option[] = [
    { title: t('dark'), value: 'dark' },
    { title: t('light'), value: 'light' },
  ];

  const languageOptions: Option[] = [
    { title: t('english'), value: 'en' },
    { title: t('german'), value: 'de' },
  ];

  const handleChangeLanguage = (lang: string) => {
    i18n.changeLanguage(lang, () => window.location.reload());
  };

  const passwordRef = useRef<HTMLInputElement>(null);
  const passwordRepeatRef = useRef<HTMLInputElement>(null);

  const passwordChangeValidationSchema = Yup.object().shape({
    password: Yup.string()
      .required(t('input_required') || '')
      .min(8, t('password_min_chars_error') || ''),
    passwordRepeat: Yup.string()
      .required(t('passwords_do_not_match') || '')
      .oneOf([Yup.ref('password')], t('passwords_do_not_match') || ''),
  });

  const passwordChangeFormik = useFormik({
    initialValues: {
      password: '',
      passwordRepeat: '',
    },
    validationSchema: passwordChangeValidationSchema,
    onSubmit: async ({ password }) => {
      await update({ password }, t('password_updated_successfully'));

      passwordChangeFormik.resetForm();
    },
  });

  useFormikErrorFocus(passwordChangeFormik, passwordRef, passwordRepeatRef);

  const usernameRef = useRef<HTMLInputElement>(null);
  const emailRef = useRef<HTMLInputElement>(null);
  const displayNameRef = useRef<HTMLInputElement>(null);

  const userSettingsValidationSchema = Yup.object().shape({
    username: Yup.string().required(t('input_required') || ''),
    email: Yup.string().required(t('input_required') || ''),
    displayName: Yup.string().required(t('input_required') || ''),
  });

  const userSettingsFormik = useFormik({
    initialValues: {
      username: '',
      email: '',
      displayName: '',
    },
    validationSchema: userSettingsValidationSchema,
    onSubmit: async ({ username, email, displayName }) => {
      await update({ username, email, displayName });
    },
  });

  useFormikErrorFocus(
    userSettingsFormik,
    usernameRef,
    emailRef,
    displayNameRef
  );

  useEffect(() => {
    if (user) {
      userSettingsFormik.setValues({
        username: user.username,
        email: user.email,
        displayName: user.displayName,
      });
    }
  }, [user]);

  return (
    <div className={styles.Settings}>
      <div className={styles.heading}>
        <h1>{t('settings')}</h1>
      </div>
      <div className={styles.row}>
        <Form
          onSubmit={userSettingsFormik.handleSubmit}
          className={styles.section}
        >
          <Input
            label={t('username')}
            placeholder={t('username')}
            value={userSettingsFormik.values.username}
            name="username"
            onChange={userSettingsFormik.handleChange}
            onBlur={userSettingsFormik.handleBlur}
            error={
              userSettingsFormik.touched.username
                ? userSettingsFormik.errors.username
                : undefined
            }
            ref={usernameRef}
            fullWidth
          />
          <Input
            label={t('email')}
            placeholder={t('email')}
            value={userSettingsFormik.values.email}
            name="email"
            onChange={userSettingsFormik.handleChange}
            onBlur={userSettingsFormik.handleBlur}
            error={
              userSettingsFormik.touched.email
                ? userSettingsFormik.errors.email
                : undefined
            }
            ref={emailRef}
            fullWidth
          />
          <Input
            label={t('display_name')}
            placeholder={t('display_name')}
            value={userSettingsFormik.values.displayName}
            name="displayName"
            onChange={userSettingsFormik.handleChange}
            onBlur={userSettingsFormik.handleBlur}
            error={
              userSettingsFormik.touched.displayName
                ? userSettingsFormik.errors.displayName
                : undefined
            }
            ref={displayNameRef}
            fullWidth
          />
          <div>
            <Button type="submit" label={t('save')} />
          </div>
        </Form>
      </div>
      <div className={styles.row}>
        <Form
          onSubmit={passwordChangeFormik.handleSubmit}
          className={styles.section}
        >
          <Input
            placeholder={t('password')}
            label={t('password')}
            value={passwordChangeFormik.values.password}
            name="password"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.password
                ? passwordChangeFormik.errors.password
                : undefined
            }
            ref={passwordRef}
            fullWidth
          />
          <Input
            label={t('password_repeat')}
            placeholder={t('password_repeat')}
            value={passwordChangeFormik.values.passwordRepeat}
            name="passwordRepeat"
            type="password"
            onChange={passwordChangeFormik.handleChange}
            onBlur={passwordChangeFormik.handleBlur}
            error={
              passwordChangeFormik.touched.passwordRepeat
                ? passwordChangeFormik.errors.passwordRepeat
                : undefined
            }
            ref={passwordRepeatRef}
            fullWidth
          />
          <div>
            <Button type="submit" label={t('save')} />
          </div>
        </Form>
      </div>
      <div className={styles.row}>
        <div className={styles.section}>
          <Dropdown
            label={t('appearance')}
            value={theme}
            options={themeOptions}
            onChange={(theme) => setTheme(theme as Theme)}
            fullWidth
          />
          <Dropdown
            label={t('language')}
            value={i18n.language}
            options={languageOptions}
            onChange={handleChangeLanguage}
            fullWidth
          />
          <div>
            <Button
              label={t('logout')}
              style="red"
              onClick={() => navigate('/logout')}
            />
          </div>
        </div>
      </div>
    </div>
  );
}

export default SettingsView;
