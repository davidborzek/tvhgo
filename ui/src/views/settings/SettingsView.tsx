import { useTranslation } from 'react-i18next';
import { Outlet } from 'react-router-dom';

import TabNavigation from '@/components/common/tabs/TabNavigation';

import styles from './SettingsView.module.scss';

export function Component() {
  const { t } = useTranslation();

  return (
    <div className={styles.Settings}>
      <div className={styles.heading}>
        <TabNavigation
          tabs={[
            {
              title: t('general'),
              to: '/settings/general',
            },
            {
              title: t('security'),
              to: '/settings/security',
            },
          ]}
        />
      </div>
      <div className={styles.content}>
        <Outlet />
      </div>
    </div>
  );
}

Component.displayName = 'SettingsView';
