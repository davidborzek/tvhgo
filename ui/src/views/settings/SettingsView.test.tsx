import { cleanup, render } from '@testing-library/react';
import { afterEach, expect, test, vi } from 'vitest';
import SettingsView from './SettingsView';
import { BrowserRouter } from 'react-router-dom';

afterEach(() => {
  vi.restoreAllMocks();
  cleanup();
});

test('should render', () => {
  const document = render(<SettingsView />, { wrapper: BrowserRouter });
  expect(document.asFragment()).toMatchSnapshot();
});
