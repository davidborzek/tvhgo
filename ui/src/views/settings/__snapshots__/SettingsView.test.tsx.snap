// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<DocumentFragment>
  <div
    class="_Settings_65046f"
  >
    <div
      class="_heading_65046f"
    >
      <nav
        class="_Tabs_593fdd"
      >
        <div
          class="_TabItem_593fdd "
        >
          <a
            class="_button_593fdd "
            href="/settings/general"
          >
            general
          </a>
        </div>
        <div
          class="_TabItem_593fdd "
        >
          <a
            class="_button_593fdd "
            href="/settings/security"
          >
            security
          </a>
        </div>
      </nav>
    </div>
    <div
      class="_content_65046f"
    />
  </div>
</DocumentFragment>
`;
